### DISCUSSION TITLE
A Store for EWW based Widgets

### DISCUSSION THREAD

**Discussion by Mesum-Hussain** [elkowar/eww#948] [Ideas]
I was thinking what if someone made a Repo of all the EWWs and made a Store app to access them.
On Gnome for example, we could make an Extension to configure EWW, and then use the Store to install Widgets..... basically Extending Gnome to add support for Widgets.
This Store could also be used on any other DE/WM to install EWWs but you need to set up EWW on your own.

**Comment by max-ishere** [comment]
A package manager would be cool. Due to the fact that eww allows `.` (dots) and `::` (rust path access) you can make a namespace for your package/module/collection and not have name collisions.

**Comment by fishhstyx** [comment]
> I was thinking what if someone made a Repo of all the EWWs and made a Store app to access them.
> On Gnome for example, we could make an Extension to configure EWW, and then use the Store to install Widgets..... basically Extending Gnome to add support for Widgets.
> This Store could also be used on any other DE/WM to install EWWs but you need to set up EWW on your own.

has this been done yet to your knowledge?

**Reply by Mesum-Hussain** [reply]
No we have better tools than EWW now so it's community is basically shrinking. 

Biggest compititor is AGS. Does the same thing but instead of Yack, you write your code in JSX (React like code). 

I hope someone makes a store for AGS widgets at least.

**Comment by battlesquid** [comment]
It always struck me as odd that EWW never leaned towards self-contained, sharable widgets; my assumption that it was is what drew me to learning it. I think it's definitely possible. That said, I personally don't think stores are the solution, I think they'd be a byproduct of solving what I think is a bigger underlying issue: standardization.

There's a distinct lack of standardization in certain areas that make it hard to readily share components. For example, there's no standard way to declare widget dependencies. One person's widget could require python, another zsh, etc etc. There's the problem of managing these dependencies across different platforms as well. A dependency might not even be a program, it might be something as simple as an API key or environment variable, and in EWW's current state, there's no standard way to supply/consume those. There's also no standard library for EWW like there is with AGS (astal), which would probably alleviate the external dep resolution problem.

I think it's probably just a matter of figuring out how to standardize things like that. If that gets done, then it becomes a matter of pulling widgets akin to how lazy.nvim pulls plugins, which is much nicer IMO because there's no single source you're dependent on for widgets.

That's my 2 cents on this matter.

**Reply by elkowar** [reply]
I'll preface this by saying that I'm not actually opposed to some sort of widget library / store, and further standardisation would be a nice upgrade.
I'll also say that I've had very little time / motivation to work on eww, so take my opinions with a grain of salt ^^
that said:

I think part of why I never bothered doing so, is that the goal for eww, moreso than for any of the newer alternatives, was to motivate people to do their own thing from scratch. I think over the first few years of eww's lifespan, that goal was actually achieved quite well: Most people made their setups from the ground up. they where often janky, hard to share / reuse, but they _did_ end up with a lot of people exploring interesting design and UX concepts. AGS, Quickshell and the like, put way more focus on sharability, reusability, and other convenience features for sharing setups. This is obviously very nice for the large set of users that mainly want to get some cool looking setup from github / reddit, do a bit of customization, and be done with it. Especially with quickshell, we're now seeing that full shell setups like noctalia and dms are taking off. It's clear that that's something the userbase wants.

However, I'm not sure if it's in the "spirit" of eww to focus on that. Unlike quickshell and ags, eww _does_ try to be a bit simpler to get into. simple configuration language, no need to learn a full programming language. (of course, for many devs, that's actually a drawback, but I'd argue eww kind of "tricked" many non-dev people into having the courage to try to build their own setups, by calling yuck a _config_ language rather than a scripting langauge.). In that sense, eww encourages starting from scratch, which often ends with more creativity, and a setup that's more specifically catered to the users needs and aesthetic sensibilities.



On the other hand: I fully agree that some sort of baseline system for declaring dependencies and wrapping scripts would be nice. While UI-components are fun to customize, I don't think anyone is hyped about writing the 100th shellscript to fetch the current wifi signal strength in some way. Having more thorough, reusable modules for these sorts of things would definitely be a massive upgrade.
